conversion_proxy/src/lib.rs

```diff
@@ -310,6 +310,35 @@ impl ConversionProxy {
         return 0_u128;
     }
 
+    /// Recursive util to convert an `amount` with a `converion_rate` having `decimals`.
+    /// The `precision` is used to work around overflows, a precision of 10^n means that the result has a precision of n yocto digits.
+    /// Said another way, a precision of 1000 will give a result rounded to the closest 1000 yoctos, ending with "...000" in yocto.
+    #[private]
+    pub fn apply_conversion_with_precision(
+        amount: U128,
+        decimals: u32,
+        conversion_rate: u128,
+        precision: u128,
+    ) -> u128 {
+        let (main_payment, flag) = (Balance::from(amount) * ONE_NEAR / ONE_FIAT / precision)
+            .overflowing_mul(10u128.pow(u32::from(decimals)));
+        let main_payment = main_payment / conversion_rate * precision;
```

Comment:
I'm not sure if this is like Solidity, but shouldn't we apply all multiplications before applying divisions?

```suggestion
        let main_payment = main_payment * precision / conversion_rate;
```

Reply:
💯

Reply:
I realize now that the tests fail that I'll have to rollback this change, it defeats the point. The multiplication should be done as as last step to not overflow (it was crafted to let the multiplication just before not overflow, so if we multiply before dividing, it will overflow)